File: stlfed_py/stlfed_py/io/get.py
import requests
import datetime # stdlib
import tempfile #stlib
import pandas as pd

# get_fred() defaults
today = datetime.datetime.now().strftime('%Y-%m-%d')
start_date_default = (pd.to_datetime(today) - pd.DateOffset(years = 5)).strftime('%Y-%m-%d')
end_date_default = today

def get_fred(fred_id, start_date = start_date_default, end_date = end_date_default):
    """
    Fetch time-series data from Federal Reserve Economic Data (FRED), provided
    by the Federal Reserve Bank of St. Louis.

    :param fred_id: ID string for FRED series

    :param start_date: Start date for FRED series, as a string in the format
                       YYYY-MM-DD. Defaults to five years prior to the date
                       called.

    :param end_date: End date for FRED series. Defaults to the date called.

    :returns: A pandas `DataFrame`
    """
    # Shitty way around not asking for an API key, since returned data is raw
    # text. With a key, change this to GET from the API endpoint
    try:
        [datetime.datetime.strptime(d, '%Y-%m-%d') for d in [start_date, end_date]]
    except ValueError:
        raise ValueError('Bad date format! Must be YYYY-MM-DD')
    
    url = 'https://fred.stlouisfed.org/graph/fredgraph.csv'
    r = requests.get(f'{url}?id={fred_id}&cosd={start_date}&coed={end_date}')

    with tempfile.NamedTemporaryFile('w', delete = False) as f:
        f.write(r.text)
        tmpfile = f.name

    df = pd.read_csv(tmpfile)

    # Prep
    df['DATE'] = pd.to_datetime(df['DATE'])
    df['label'] = 'Actual'

    return df
# end get_fred

File: stlfedR/R/get.R
#' Get FRED Data into DF
#' 
#' See title
#'
#' @param fred_id ID for FRED data ticker
#' @param years_history NUmber of hostoirial years of data to grab who car about splenbign
#' @export
get_FRED <- function(fred_id, years_history) {
    url <- 'https://fred.stlouisfed.org/graph/fredgraph.csv'
    today <- Sys.Date()
    start_date <- today - years(years_history)
    end_date <- today
    
    # Shitty way around not asking for an API key
    # Should eventually call httr::GET() with a key, to directly return a dataframe
    tmpfile <- tempfile(fileext = '.csv')
    download.file(glue::glue('{url}?id={fred_id}&cosd={start_date}&coed={end_date}'), tmpfile)
    
    df_0 <- read.csv(tmpfile, strip.white = TRUE, stringsAsFactors = FALSE)
    df_0$DATE <- as_date(df_0$DATE)
    df_0$label <- 'Actual'
    
    return(df_0)
}
